### ISSUE TITLE
Light control is working for me, for whatever reason fan control does not.

### ISSUE THREAD

**Issue by xNUTx** [xNUTx/esphome-components#3]
@Xmister I finally had a chance to sit down and test your alterations.

First of all, a BIG, BIG thank you for the work you did. I really appreciate it :+1: 

I ran into the issue that light control works fine, as does pairing the unit with my test board. Fan control on the other hand still seems to be a little illusive. The fan does not turn on and the unit usually beeps when you do, with the remote, but now it stayed silent and nothing happened.

I tried to figure out how to fix this but truthfully, my C(++) programming skills are woefully underdeveloped. Do you have any suggestions what I can do to properly test the functionality on my fans?

**Comment by xNUTx** [comment]
`[20:30:56][C][lampsmartpro:140]: LampSmartProLight 'Huiskamer Light'
[20:30:56][C][lampsmartpro:141]:   Cold White Temperature: 167.000000 mireds
[20:30:56][C][lampsmartpro:142]:   Warm White Temperature: 333.000000 mireds
[20:30:56][C][lampsmartpro:143]:   Constant Brightness: false
[20:30:56][C][lampsmartpro:144]:   Minimum Brightness: 7
[20:30:56][C][lampsmartpro:145]:   Transmission Duratoin: 750 millis
[20:30:56][C][fansmartpro:122]: FanSmartProFan ''`
I am wondering if it has to do with this... it does not relay the name set for the device in the logs. It does however show up in HA.

**Comment by Xmister** [comment]
@xNUTx Well I just copied my remote signal verbatim, so it's possible that the code does not work for every lamp.
BTW. This is how I did it:
Esphome code:
```
esp32_ble_tracker:
  scan_parameters:
    active: False
    continuous: False
    interval: 1000ms
    window: 1000ms
  on_ble_advertise: 
      mac_address:
        - 59:BA:2B:D2:57:8D
        - 10:9E:3A:10:25:5D
        - 0c:90:43:19:e2:c8
      then:
        - lambda: |-
            ESP_LOGD("ble_adv", "New BLE device");
            ESP_LOGD("ble_adv", "  address: %s", x.address_str().c_str());
            ESP_LOGD("ble_adv", "  name: %s", x.get_name().c_str());
            ESP_LOGD("ble_adv", "  Advertised service UUIDs:");
            for (auto uuid : x.get_service_uuids()) {
                ESP_LOGD("ble_adv", "    - %s", uuid.to_string().c_str());
            }
            ESP_LOGD("ble_adv", "  Advertised service data:");
            for (auto data : x.get_service_datas()) {
                char ss[data.data.size()*6 + 4];
                sprintf(ss, "{ ");
                char* sspos = &ss[2];
                for (auto d: data.data) {
                  sprintf(sspos,"0x%02X, ",d);
                  sspos+=6;
                }
                sprintf(sspos-2," }");
                ESP_LOGD("ble_adv", "    - %s: %s", data.uuid.to_string().c_str(), ss);
            }
            ESP_LOGD("ble_adv", "  Advertised manufacturer data:");
            char num[4] = {0};
            for (auto data : x.get_manufacturer_datas()) {
                char ss[data.data.size()*6 + 4];
                sprintf(ss, "{ ");
                char* sspos = &ss[2];
                for (auto d: data.data) {
                  sprintf(sspos,"0x%02X, ",d);
                  sspos+=6;
                }
                sprintf(sspos-2," }");
                ESP_LOGD("ble_adv", "    - %s: %s", data.uuid.to_string().c_str(), ss);

            }
```
The listed MACs are the ones my remote used, you may need to remove them to get all the data, and once you identify your remote, you can filter for those.
Then you take the resulting service UUID and data for you buttons and replace the values in fan.h

**Comment by xNUTx** [comment]
Ah! That makes perfect sense now. The way you did it will only work for your lamp and not ever will work for any other lamp... I had been looking at your commands you issue, and saw the way you did it is so different from the lamp controls I expected something alike to be the case...

The thing is, between the remote and the lamp there has been a pairing. This pairing generates a code/serial number and that is used to encode the communications to make sure the signal used is legitimate and safe to execute... very required if you have more then one fan/lamp in the house working on BLE with the same type of remote...

If you look over the lighting controls you can see it builds up the desired command and then encrypts it before sending it.

I wanted to have the fan part working the same way, so it would be usable for everyone with a similar fan-light.

**Comment by Xmister** [comment]
Yeah,  initially I wanted to do the same, but it just didn't want to work. Maybe the encryption was different, but I might have another try at it if I get some time.
You can also try to capture fan command codes properly. You can use the esphome code, then the unwhit.py to get the actual code. If you get the same result for multiple tries for the same function, we can make the code work as the lamp one.

**Comment by xNUTx** [comment]
This is the original source for control codes from the guy who made the initial version with light control only:

https://gist.github.com/aronsky/f433de654f008fedb5161e08eb32c33e

And the java code using them:

https://gist.github.com/aronsky/f2d8afab134d15f34256187a82a53a9c

Obviously, you can imagine I can read it a bit, but besides some javascript... this is not my strong suit either...

**Comment by xNUTx** [comment]
@Xmister do you by chance still have your initial code attempts while you where still working on trying to get the paired mode to work? My knowledge may be limited, but it will be a starting point I would think.

I have 2 of these fans, I really do not like to use a code base for both of them, I would really prefer the 'official' way if possible (and we know it is, as you can see the way it works on the light portion).

**Comment by Xmister** [comment]
@xNUTx I have a code locally that works with the fan with the same kind of generated commands as light does. I'm still trying to find all the codes, but I'll upload it soon.

**Comment by xNUTx** [comment]
The codes are known, from the java code you can see what it expects together with the command code as parameters. Some values are simply booleans. I also worked on it myself reintroducing some code that got removed. Maybe you can use some of it, though I would not trust my handy work over that of your own. Either way, keep up the good work!! 

Week holiday from today... I won't be working on anything during that time for sure.

⁣--

Met vriendelijke groeten/Regards,

Barry Schut
a.k.a. [NUT]​

Op 17 jul. 2024 22:42, om 22:42, "Péter Sárközi" ***@***.***> schreef:

**Comment by drudgebg** [comment]
[NicoIIT](https://github.com/NicoIIT)
/
[esphome-components](https://github.com/NicoIIT/esphome-components)
Hello, I tested this fork today and I can say it works with my Celling fan with led.

**Comment by xNUTx** [comment]
I'll be testing it next week 😁

⁣--

Met vriendelijke groeten/Regards,

Barry Schut
a.k.a. [NUT]​

Op 18 jul. 2024 11:23, om 11:23, drudgebg ***@***.***> schreef:

**Comment by xNUTx** [comment]
I am going to retire our work @Xmister as it seems @NicoIIT made it work. It has so many improvements I am going to sync up to his and drop everyting we both did.

**Comment by Xmister** [comment]
Yeah, I also rebased on that, but added state restoration for fans.
